indexer/Worker.cc

```diff
@@ -36,8 +54,424 @@ struct MessageQueuePair {
+  HashValue runningHash;
+  // Optional field to track all the inputs that went into computing
+  // a hash, meant for debugging.
+  std::unique_ptr<std::vector<std::string>> history;
+
+public:
+  HashValueBuilder(bool recordHistory)
+      : runningHash(),
+        history(recordHistory ? std::make_unique<std::vector<std::string>>()
+                              : nullptr) {}
+
+  void mix(llvm::StringRef text) {
+    this->runningHash.mix(reinterpret_cast<const uint8_t *>(text.data()),
+                          text.size());
+    if (this->history) {
+      this->history->emplace_back(text.str());
+    }
+  }
+
+  void mix(uint64_t v) {
+    this->runningHash.mix(reinterpret_cast<const uint8_t *>(&v), sizeof(v));
+    if (this->history) {
+      this->history->emplace_back(fmt::format("{}", v));
+    }
+  }
+
+  void mixString(std::string &&text) {
+    this->mix(llvm::StringRef(text));
+  }
+
+  HashValue finish() const {
+    return this->runningHash;
+  }
+
+  bool isRecordingHistory() {
+    return (bool)this->history;
+  }
+};
+
+#define MIX_WITH_KEY(__hash, __key_expr, __value)                   \
+  {                                                                 \
+    if (__hash.isRecordingHistory()) {                              \
+      __hash.mixString(fmt::format("{}: {}", __key_expr, __value)); \
+    } else {                                                        \
+      __hash.mix(__value);                                          \
+    }                                                               \
+  }
+
+struct HeaderInfoBuilder final {
+  const clang::FileID fileId;
+  HashValueBuilder hashValueBuilder;
+};
+
+struct MultiHashValue {
+  const HashValue hashValue;
+  bool isMultiple;
+};
+
+class IndexerPPStack final {
+public:
+  std::vector<HeaderInfoBuilder> state;
+  // Headers which we've seen only expand in a single way.
+  // The extra bit inside the MultiHashValue struct indicates
+  // if should look in the finishedProcessingMulti map instead.
+  absl::flat_hash_map<LLVMToAbslHashAdapter<clang::FileID>, MultiHashValue>
+      finishedProcessing;
+  // Headers which expand in at least 2 different ways.
+  // The values have size() >= 2.
+  absl::flat_hash_map<LLVMToAbslHashAdapter<clang::FileID>,
+                      absl::flat_hash_set<HashValue>>
+      finishedProcessingMulti;
+
+public:
+  bool empty() {
+    return this->state.empty();
+  }
+  HashValueBuilder &topHash() {
+    ENFORCE(!this->empty());
+    return this->state.back().hashValueBuilder;
+  }
+
+  void enterInclude(clang::FileID fileId, bool recordHistory) {
+    this->state.emplace_back(
+        HeaderInfoBuilder{fileId, HashValueBuilder(recordHistory)});
+  }
+  std::optional<HashValue>
+  exitInclude(clang::FileID fileId, const clang::SourceManager &sourceManager) {
+    if (fileId.isInvalid()) {
+      return {};
+    }
+    ENFORCE(!this->state.empty(), "missing matching enterInclude for exit");
```

Comment:
Q: Right now, the core logic is kinda' spread out across this type and the IndexerPPCallbacks type. Should we move all the logic into the callbacks type? That also has access to the SourceManager, which makes debugging a little easier.